src/aap_eda/tasks/ruleset.py

```diff
@@ -29,34 +30,98 @@
 
 
 @job("activation")
-def activate_rulesets(is_restart: bool, activation_id: int) -> None:
-    activation = models.Activation.objects.filter(id=activation_id).first()
-    if (
-        not activation
-        or str(activation.status) == ActivationStatus.DELETING.value
-    ):
-        logger.info(f"Activation id: {activation_id} is deleted")
-        return
+def activate(activation_id: int, requester: str = "User") -> None:
+    with transaction.atomic():
+        activation = (
+            models.Activation.objects.select_for_update()
+            .filter(id=activation_id)
+            .first()
+        )
+        if (
+            not activation
+            or str(activation.status) == ActivationStatus.DELETING.value
+        ):
+            logger.info(f"Activation id: {activation_id} is deleted")
+            return
 
-    # TODO(hsong): move this to views
-    if is_restart:
-        activation.restart_count += 1
-        activation.save(update_fields=["restart_count", "modified_at"])
-    logger.info(f"Task started: Activate rulesets ({activation.name})")
+        logger.info(
+            f"Activating activation id: {activation_id} requested "
+            f"by {requester}"
+        )
+
+        activation.status = ActivationStatus.STARTING
+        activation.save(update_fields=["status"])
 
     if activation.is_enabled:
+        # Note: expect scheduler to set requester as "SCHEDULER"
+        if requester == "SCHEDULER":
+            activation.restart_count += 1
+            activation.save(update_fields=["restart_count", "modified_at"])
+
         ActivateRulesets().activate(activation)
 
-        logger.info(f"Task finished: Rulesets ({activation.name}).")
+        logger.info(
+            f"Task finished: activation {activation.name}"
+            f" {activation.status}."
+        )
     else:
         logger.info(
-            f"Task finished: Rulesets ({activation.name}) has been disabled."
+            f"Task finished: activation id {activation_id} has been disabled."
         )
 
 
 @job("default")
-def deactivate(activation_id: int, is_delete: bool = False) -> None:
-    activation = models.Activation.objects.get(id=activation_id)
+def deactivate(
+    activation_id: int, requester: str = "User", delete: bool = False
+) -> None:
+    logger.info(
+        f"Disabling activation id: {activation_id} requested by {requester}"
+    )
+
+    with transaction.atomic():
+        activation = (
+            models.Activation.objects.select_for_update()
+            .filter(id=activation_id)
+            .first()
+        )
+        if not activation:
+            logger.warning(f"Activation {activation_id} is deleted.")
+            return
+
+        _perform_deactivate(activation)
```

Comment:
Before this line we should validate activation status is in pending, running, starting, or unresponsive. Otherwise log and return.